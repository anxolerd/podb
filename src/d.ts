declare module 'flora-sql-parser';
declare module 'gettext-parser';