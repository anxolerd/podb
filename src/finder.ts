function trim(string: string): string {
  return string.replace(/^\s+|\s+$/g, "");
}

export function find(data: string, msgidToFind: string, msgctxtToFind: string) {
  //support both unix and windows newline formats.
  data = data.replace(/\r\n/g, "\n");
  var lineNo = 0;
  var lineNoStart = 0;
  var lineNoFinish = 0;
  var context = null;
  var msgid = "";
  var msgctxt = "";

  function finish() {
    if (msgid.length > 0) {
      if (msgid == msgidToFind && msgctxt == msgctxtToFind) {
        lineNoFinish = lineNo - 1;
      } else {
        msgid = "";
        msgctxt = "";
        lineNoStart = lineNo;
      }
    }
  }

  function extract(string: string) {
    string = trim(string);
    string = string.replace(/^[^"]*"|"$/g, "");
    string = string.replace(
      /\\([abtnvfr'"\\?]|([0-7]{3})|x([0-9a-fA-F]{2}))/g,
      function(match, esc, oct, hex) {
        if (oct) {
          return String.fromCharCode(parseInt(oct, 8));
        }
        if (hex) {
          return String.fromCharCode(parseInt(hex, 16));
        }
        switch (esc) {
          case "a":
            return "\x07";
          case "b":
            return "\b";
          case "t":
            return "\t";
          case "n":
            return "\n";
          case "v":
            return "\v";
          case "f":
            return "\f";
          case "r":
            return "\r";
          default:
            return esc;
        }
      }
    );
    return string;
  }
  const lines = data.split("\n");
  while (lines.length > 0) {
    var line = trim(<string>lines.shift());
    lineNo += 1;
    if (line.match(/^#:/)) {
      // Reference
      finish();
    } else if (line.match(/^#,/)) {
      // Flags
      finish();
    } else if (line.match(/^#($|\s+)/)) {
      // Translator comment
      finish();
    } else if (line.match(/^#\./)) {
      // Extracted comment
      finish();
    } else if (line.match(/^msgid_plural/)) {
      // Plural form
      context = "msgid_plural";
    } else if (line.match(/^msgid/)) {
      // Original
      finish();
      msgid = extract(line);
      context = "msgid";
    } else if (line.match(/^msgstr/)) {
      // Translation
      if (msgid == "") {
        // Header hack-fix
        msgid = "___________________dummy";
      }
      context = "msgstr";
    } else if (line.match(/^msgctxt/)) {
      // Context
      finish();
      msgctxt = extract(line);
      context = "msgctxt";
    } else {
      // Probably multiline string or blank
      if (line.length > 0) {
        if (context === "msgid") {
          msgid += extract(line);
        } else if (context === "msgctxt") {
          msgctxt += extract(line);
        }
      }
    }
    if (lineNoFinish) {
      return [lineNoStart, lineNoFinish];
    }
  }
  finish();
  return [-1, -1];
}
