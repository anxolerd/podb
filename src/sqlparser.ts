import { Parser } from "flora-sql-parser";

const p = new Parser();

export enum QueryType {
  SELECT = "select",
  UPDATE = "update"
}

export enum OperatorType {
  AND = "AND",
  NOT = "NOT",
  EQUAL = "="
}

export enum WhereType {
  UNARY = "unary_expr",
  COLUMN = "column_ref",
  BINARY = "binary_expr"
}

export type FromClause = {
  table: string;
};

export type ColumnRef = {
  column: string;
  value: string;
  type: string;
  table?: string;
};

export type WhereClause = {
  operator: OperatorType;
  left: ColumnRef | WhereClause;
  right: ColumnRef | WhereClause;
  expr: WhereClause | ColumnRef;
  type: string;
};

export type Expression = {
  expr: ColumnRef;
};

export type SetValue = {
  column: string;
  value: { type: string; value: string };
};

export type ParsedStatement = {
  type: string;
  where: WhereClause;
  from: Array<FromClause>;
  columns: Array<Expression>;
  set: Array<SetValue>;
  table: string;
};

export function parseSql(statement: string): ParsedStatement {
  return p.parse(statement);
}

// > ss = p.parse('SELECT * FROM t where msgstr="ddsa"');
// { with: null,
//   type: 'select',
//   options: null,
//   distinct: null,
//   columns: '*',
//   from: [ { db: null, table: 't', as: null } ],
//   where:
//    { type: 'binary_expr',
//      operator: '=',
//      left: { type: 'column_ref', table: null, column: 'msgstr' },
//      right: { type: 'column_ref', table: null, column: 'ddsa' } },
//   groupby: null,
//   having: null,
//   orderby: null,
//   limit: null }
// > ss.where
// { type: 'binary_expr',
//   operator: '=',
//   left: { type: 'column_ref', table: null, column: 'msgstr' },
//   right: { type: 'column_ref', table: null, column: 'ddsa' } }
