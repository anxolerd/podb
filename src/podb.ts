import * as fs from "fs";
import { find } from "./finder";
import { Item, parse, PO } from "pofile";
import {
  parseSql,
  WhereClause,
  WhereType,
  OperatorType,
  ColumnRef,
  SetValue,
  QueryType,
  ParsedStatement
} from "./sqlparser";

const AVAILABLE_COLUMNS = ["msgid", "msgstr", "msgctxt", "msgid_plural"];
const AVAILABLE_COLUMNS_EXCEPTION = `Available columns: ${AVAILABLE_COLUMNS.join(
  ", "
)}`;

export class PoTable {
  po: PO;
  rawData: string;

  constructor(tableData: string) {
    this.rawData = tableData;
    this.po = parse(tableData);
  }

  _match(item: Item, where: WhereClause | ColumnRef): boolean {
    if (where.type == WhereType.COLUMN) {
      const condition = <ColumnRef>where;
      if (AVAILABLE_COLUMNS.indexOf(condition.column) == -1) {
        throw AVAILABLE_COLUMNS_EXCEPTION;
      }
      if (condition.column == "msgid") return !!item.msgid;
      if (condition.column == "msgstr") return !!item.msgstr[0];
      if (condition.column == "msgctxt") return !!item.msgctxt;
      if (condition.column == "msgid_plural") return !!item.msgid_plural;
    } else {
      where = <WhereClause>where;
      if (where.operator == OperatorType.EQUAL) {
        const condition = <ColumnRef>where.left;
        const value = (<ColumnRef>where.right).value;
        if (condition.column == "msgid") {
          return item.msgid == value;
        } else if (condition.column == "msgstr") {
          return item.msgstr[0] == value;
        } else if (condition.column == "msgctxt") {
          return (item.msgctxt || "") == value;
        } else {
          throw AVAILABLE_COLUMNS_EXCEPTION;
        }
      } else if (where.operator == OperatorType.AND) {
        return (
          this._match(item, <WhereClause>where.left) &&
          this._match(item, <WhereClause>where.right)
        );
      } else if (where.operator == OperatorType.NOT) {
        return !this._match(item, <WhereClause>where.expr);
      }
    }
    return false;
  }

  _set(item: Item, setValues: Array<SetValue>) {
    for (const setValue of setValues) {
      switch (setValue.column) {
        case "msgid": {
          item.msgid = setValue.value.value;
          continue;
        }
        case "msgstr": {
          item.msgstr = [setValue.value.value];
        }
        case "msgctxt": {
          item.msgctxt = setValue.value.value;
        }
      }
    }
  }

  _select(items: Array<Item>, where: WhereClause) {
    const result = [];
    for (const item of items) {
      if (this._match(item, where)) {
        result.push(item);
      }
    }
    return result;
  }

  _update(
    items: Array<Item>,
    where: WhereClause,
    setValues: Array<SetValue>
  ): number {
    let updatedCount = 0;
    for (const item of items) {
      if (this._match(item, where)) {
        const msgid = item.msgid;
        const msgctxt = item.msgctxt || "";
        this._set(item, setValues);
        const [start, finish] = find(this.rawData, msgid, msgctxt);
        const splitData = this.rawData.split("\n");
        this.rawData = splitData
          .slice(0, start - 1)
          .concat([item.toString(), ""])
          .concat(splitData.slice(finish))
          .join("\n");
        updatedCount += 1;
      }
    }
    this.po = parse(this.rawData);
    return updatedCount;
  }

  execute(statement: string): number | Array<Item> {
    const sql = parseSql(statement);
    switch (sql.type) {
      case QueryType.SELECT: {
        return this._select(this.po.items, sql.where);
      }
      case QueryType.UPDATE: {
        return this._update(this.po.items, sql.where, sql.set);
      }
      default:
        throw "Only selects and updates are supported";
    }
  }
}

export class PoDb {
  path: string;

  constructor(path: string) {
    this.path = path;
  }

  _getFilename(sql: ParsedStatement) {
    switch (sql.type) {
      case QueryType.SELECT:
        return sql.from[0].table;
      case QueryType.UPDATE:
        return sql.table;
      default:
        throw "Only selects and updates are supported";
    }
  }

  execute(statement: string, sync: Boolean): number | Array<Item> {
    const sql = parseSql(statement);
    const filename = this._getFilename(sql);
    const path = `${this.path}/${filename}.po`;
    let data = fs.readFileSync(path).toString();
    const poTable = new PoTable(data);
    const result = poTable.execute(statement);
    if (sync) {
      fs.writeFileSync(path, poTable.rawData);
    }
    return result;
  }
}
