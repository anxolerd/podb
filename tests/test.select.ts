import * as fs from "fs";
import * as path from "path";
import { PoTable } from "../src/podb";

test("select * from table", () => {
  const data = fs
    .readFileSync(path.resolve(__dirname, "fixtures/stats.po"))
    .toString();
  const table = new PoTable(data);
  expect(table.execute("select * from t where msgid")).toMatchSnapshot();
});
