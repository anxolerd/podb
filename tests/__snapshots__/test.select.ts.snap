// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select * from table 1`] = `
Array [
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": null,
    "msgid": "test",
    "msgid_plural": null,
    "msgstr": Array [
      "",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [
      "tests/fixtures/checkTest/check-trans-exist.js:2",
      "tests/fixtures/checkTest/check-trans-exist.js:3",
    ],
  },
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": "bla",
    "msgid": "test",
    "msgid_plural": null,
    "msgstr": Array [
      "",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [
      "tests/fixtures/checkTest/check-trans-exist.js:2",
      "tests/fixtures/checkTest/check-trans-exist.js:3",
    ],
  },
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": "lala",
    "msgid": "TESTUPPER",
    "msgid_plural": null,
    "msgstr": Array [
      "",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [],
  },
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": "lala",
    "msgid": "test_regexp",
    "msgid_plural": null,
    "msgstr": Array [
      "test_regeyp",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [],
  },
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": null,
    "msgid": "test_plural",
    "msgid_plural": "test_plural",
    "msgstr": Array [
      "",
      "xxx",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [],
  },
  Object {
    "comments": Array [],
    "extractedComments": Array [],
    "flags": Object {},
    "msgctxt": "bla",
    "msgid": "test_plural",
    "msgid_plural": "test_plural",
    "msgstr": Array [
      "",
    ],
    "nplurals": 2,
    "obsolete": false,
    "references": Array [],
  },
]
`;
